include/AST/Decl/FunctionDecl.hpp

```diff
@@ -10,62 +11,106 @@
 /// @brief Represents a function declaration in the AST.
 ///
 /// This class inherits from DeclBase and encapsulates the details of a function
 /// declaration, including its name, type, parameters, and body.
-class FunctionDecl : public DeclBase {
-    std::string _name;
+class FunctionDecl final
+    : public DeclBase,
+      private llvm::TrailingObjects<FunctionDecl, ParamDecl> {
+private:
+    using TrailingParams = llvm::TrailingObjects<FunctionDecl, ParamDecl>;
+    llvm::StringRef _name;
     glu::types::FunctionTy *_type;
-    llvm::SmallVector<Param> _params;
     CompoundStmt _body;
 
-public:
-    /// @brief Constructor for the FunctionDecl class.
-    /// @param location The source location of the function declaration.
-    /// @param parent The parent AST node.
-    /// @param name The name of the function.
-    /// @param type The type of the function.
-    /// @param params A vector of Param objects representing the parameters of
-    /// the function.
+    unsigned _numParams;
+
+    // Method required by llvm::TrailingObjects to determine the number
+    // of trailing objects.
+    size_t numTrailingObjects(typename TrailingParams::OverloadToken<ParamDecl>)
+        const
+    {
+        return _numParams;
+    }
+
     FunctionDecl(
-        SourceLocation location, ASTNode *parent, std::string name,
-        glu::types::FunctionTy *type, llvm::SmallVector<Param> params
+        SourceLocation location, ASTNode *parent, llvm::StringRef name,
+        glu::types::FunctionTy *type, llvm::ArrayRef<ParamDecl> const params
     )
         : DeclBase(NodeKind::FunctionDeclKind, location, parent)
         , _name(std::move(name))
         , _type(type)
-        , _params(std::move(params))
         , _body(CompoundStmt(location, {}))
+        , _numParams(params.size())
     {
+        std::uninitialized_copy(
+            params.begin(), params.end(),
+            this->template getTrailingObjects<ParamDecl>()
+        );
         _body.setParent(this);
     }
 
+public:
+    friend TrailingParams;
+
+    /// @brief Static method to create a new FunctionDecl.
+    /// @param alloc The allocator used to create the FunctionDecl.
+    /// @param location The source location of the function declaration.
+    /// @param parent The parent AST node.
+    /// @param name The name of the function.
+    /// @param type The type of the function.
+    /// @param params A vector of Param objects representing the parameters of
+    /// @return Returns a new FunctionDecl.
+    static FunctionDecl *create(
+        llvm::BumpPtrAllocator &alloc, SourceLocation location, ASTNode *parent,
+        llvm::StringRef name, glu::types::FunctionTy *type,
+        llvm::ArrayRef<ParamDecl> const params
+    )
+    {
+        auto totalSize = totalSizeToAlloc<ParamDecl>(params.size());
+        void *mem = alloc.Allocate(totalSize, alignof(FunctionDecl));
+        return new (mem)
+            FunctionDecl(location, parent, std::move(name), type, params);
```

Comment:
std::move on StringRef does nothing
```suggestion
            FunctionDecl(location, parent, name, type, params);
```